contrib/xor-blocks/src/main.rs

```diff
@@ -0,0 +1,193 @@
+                }
+            })
+    } else {
+        return Err("Invalid number of arguments. Run with `cargo run --release` if using default datadir. Otherwise run with `cargo run --release -- -datadir=<custom datadir>`.".into());
+    };
+
+    let blocks_path = data_dir_path.join("blocks");
+
+    let paths = fs::read_dir(&blocks_path)?
+        .map(|res| res.map(|e| e.path()))
+        .collect::<Result<Vec<_>, io::Error>>()?;
+
+    let key_path = blocks_path.join(XOR_FILE_NAME);
+    let key: [u8; 8] = fs::read(&key_path)
+        .map_err(|_| "No xor.dat file. Make sure you are running Bitcoin Core v28 or higher.")?
+        .try_into()
+        .map_err(|_| "Invalid xor.dat file.")?;
+
+    if key[..4] == [0u8; 4] && key[4..] != [0u8; 4] {
+        return Err(
+            "This tool doesn't work with a non-zero key with 4 bytes of leading zeros".into(),
+        );
+    }
+
+    let key = if key == [0u8; 8] {
+        // The key in the file is zero, so overwrite with a new random one
+        loop {
+            let key: [u8; 8] = RandomState::new().build_hasher().finish().to_le_bytes();
+            // Don't use keys with 4 bytes of leading zeros
+            // They won't let us detect the first 4 bytes of magic in the files
+            if key[..4] == [0u8; 4] {
+                continue;
+            }
+            let mut out_file = File::options().write(true).open(&key_path)?;
+            out_file.write_all(&key)?;
+            out_file.sync_data()?;
+            break key;
+        }
+    } else {
+        // Use the existing random key
+        key
+    };
+
+    println!("Obfuscating blocks dir. Do not start bitcoind until finished!");
+
+    let total = paths.len();
+    let mut done = 0;
+    let mut timer = Instant::now();
+    let duration = Duration::from_secs(5);
+
+    let mut double_key = [0u8; 16];
+    double_key[..8].copy_from_slice(&key);
+    double_key[8..].copy_from_slice(&key);
+    let key = u128::from_ne_bytes(double_key);
+
+    paths.into_iter().for_each(|path| {
+        if path.extension().is_none_or(|f| f != "dat") {
+            return;
+        }
+
+        if path.file_name().expect("there to be a file name") == XOR_FILE_NAME {
+            return;
+        }
+
+        if let Err(e) = xor_file(&path, key) {
+            println!("Error obfuscating file {}: {e}", path.display())
+        };
+
+        done += 1;
+        if timer.elapsed() > duration {
+            let rate = done / start.elapsed().as_secs();
+            println!("Obfuscated {done} / {total} files ({rate}/sec)");
+            timer = Instant::now();
+        }
+    });
+
+    println!(
+        "Done in {} seconds! Blocksdir is now obfuscated.",
+        start.elapsed().as_secs()
+    );
+
+    Ok(())
+}
+
+fn xor_file(path: &Path, key: u128) -> Result<(), io::Error> {
+    let mut buf_u128 = 0u128;
+    let buf = unsafe { (&mut buf_u128 as *mut _ as *mut [u8; 16]).as_mut() }.unwrap();
```

Comment:
This seems over-complicated and not needed. You're not saing anything by "avoiding the conversion" - the optimizer would take care of that for you.